src/west/cmd/project.py

```diff
@@ -0,0 +1,601 @@
+    # Raises an error if the project's clone path is not the top-level
+    # directory of a Git repository
+
+    # --is-inside-work-tree doesn't require that the directory is the top-level
+    # directory of the Git repository. Use --show-cdup instead, which prints an
+    # empty string (i.e., just a newline, which we strip) in that case.
+    res = _git(project, 'rev-parse --show-cdup', capture_stdout=True,
+               check=False)
+
+    if res.returncode != 0 or res.stdout != "":
+        _die(project, '(name-and-path) is not the top-level directory of a '
+                      'Git repository, as reported by '
+                      "'git rev-parse --show-cdup'")
+
+
+def _has_branch(project, branch):
+    return _git(project, 'show-ref --quiet --verify refs/heads/' + branch,
+                check=False).returncode == 0
+
+
+def _is_sha(s):
+    try:
+        int(s, 16)
+    except ValueError:
+        return False
+
+    return len(s) == 40
+
+
+def _git_base(project, cmd, *, extra_args=(), capture_stdout=False, check=True):
+    # Runs a git command in the West top directory. See _git_helper() for
+    # parameter documentation.
+    #
+    # Returns a CompletedProcess instance (see below).
+
+    return _git_helper(project, cmd, extra_args, util.west_topdir(),
+                       capture_stdout, check)
+
+
+def _git(project, cmd, *, extra_args=(), capture_stdout=False, check=True):
+    # Runs a git command within a particular project. See _git_helper() for
+    # parameter documentation.
+    #
+    # Returns a CompletedProcess instance (see below).
+
+    return _git_helper(project, cmd, extra_args,
+                       os.path.join(util.west_topdir(), project.path),
+                       capture_stdout, check)
+
+
+def _git_helper(project, cmd, extra_args, cwd, capture_stdout, check):
+    # Runs a git command.
+    #
+    # project:
+    #   The Project instance for the project, derived from the manifest file.
+    #
+    # cmd:
+    #   String with git arguments. Supports some "(foo)" shorthands. See below.
+    #
+    # extra_args:
+    #   List of additional arguments to pass to the git command (e.g. from the
+    #   user).
+    #
+    # cwd:
+    #   Directory to switch to first (None = current directory)
+    #
+    # capture_stdout:
+    #   True if stdout should be captured into the returned
+    #   subprocess.CompletedProcess instance instead of being printed.
+    #
+    #   We never capture stderr, to prevent error messages from being eaten.
+    #
+    # check:
+    #   True if an error should be raised if the git command finishes with a
+    #   non-zero return code.
+    #
+    # Returns a subprocess.CompletedProcess instance.
+
+    # TODO: Run once somewhere?
+    if shutil.which('git') is None:
+        log.die('Git is not installed or cannot be found')
+
+    args = [_expand_shorthands(project, arg) for arg in cmd.split()]
+
+    popen = subprocess.Popen(
+        ('git', *args, *extra_args),
```

Comment:
not python 3.4 -- mind adding some basic tests since we seem to have a pattern of using too-recent language features?